src/vsr.zig

```diff
@@ -1571,6 +1571,52 @@ pub const Checkpoint = struct {
         assert(constants.journal_slot_count > constants.lsm_batch_multiple);
         assert(constants.journal_slot_count % constants.lsm_batch_multiple == 0);
     }
+    /// Checkpoint identifier.
+    id: u128,
+    /// The op_checkpoint() that corresponds to the checkpoint id.
+    op: u64,
+
+    pub const Quorum = struct {
+        /// The latest known checkpoint identifier from every *other* replica.
+        checkpoints: [constants.replicas_max]?Checkpoint =
+            [_]?Checkpoint{null} ** constants.replicas_max,
+
+        pub fn replace(
+            quorum: *Quorum,
+            replica: u8,
+            checkpoint: *const Checkpoint,
+        ) bool {
+            if (quorum.checkpoints[replica]) |checkpoint_existing| {
+                // Ignore old candidate.
+                if (checkpoint.op < checkpoint_existing.op) {
+                    return false;
+                }
+
+                // Ignore repeat candidate.
+                if (checkpoint.op == checkpoint_existing.op and
+                    checkpoint.id == checkpoint_existing.id)
+                {
+                    return false;
```

Comment:
```suggestion
                if (checkpoint.op == checkpoint_existing.op)
                {
                    assert(checkpoint.id == checkpoint_existing.id);
                    return false;
```

Reply:
this one still stands!

Reply:
Oops, had misinterpreted this the first time you posted it and then it got lost! Added in https://github.com/tigerbeetle/tigerbeetle/pull/2104/commits/f0221b6dc846c5dfbd0fc4f61b6fbcc1857183c8. 😄